chore(sdk): clean up client.py

droctothorpe opened the pull request:
**Description of your changes:**
This PR tidies up the type hints in the client module.


**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->

connor-mccarthy commented on sdk/python/kfp/client/client.py:
```diff
@@ -1082,7 +1081,7 @@ def create_run_from_pipeline_func(
     def create_run_from_pipeline_package(
         self,
         pipeline_file: str,
-        arguments: Optional[Dict[str, str]] = None,
```
Is it possible to submit a run with `arguments={'number': 1}`?

droctothorpe replied:
I’m able to get the function to execute even when the `arguments` dictionary includes an int. [This](https://www.kubeflow.org/docs/components/pipelines/v2/author-a-pipeline/component-io/#inputs-1) documentation indicates that non-string values are an accepted pattern.

> Valid input parameter annotations include str, int, float, bool, dict, list.

connor-mccarthy replied:
In that case, perhaps the original annotation is correct?

droctothorpe replied:
Also, these two functions just wrap `run_pipeline`, which uses the `Dict[str, any]` syntax in [its type hints](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/client/client.py#L736).

`run_pipeline` invokes `_create_job_config`, which also uses [Any](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/client/client.py#L736),

droctothorpe replied:
After a few more hops, it eventually winds up at [this](https://github.com/kubeflow/pipelines/blob/master/backend/api/v1beta1/go_http_client/job_model/pipeline_spec_runtime_config.go#L21) Go struct, which has a value of `interface{}`, i.e. an empty interface, which matches _all_ types.

droctothorpe replied:
Just to clarify and add some context, [create_run_from_pipeline_func](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/client/client.py#L1027) in this _same_ module uses `Dict[str, Any]` as opposed to `Dict[str, str]`. So two functions in the same module that essentially do the same thing have different type hints. If that's not intentional, then they should probably be brought into alignment, which begs the question, which one is right? Since they both wrap around functions (and ultimately types) that use `Dict[str, Any]`, it seems like that's the standard and `create_run_from_pipeline_package` just has a small discrepancy that needs to be brought into alignment. Maybe I'm missing something though. As my wife can attest, I'm wrong all the time.

connor-mccarthy replied:
Thanks for explaining. I misunderstood :). This change LGTM!

connor-mccarthy replied:
I believe the TODO comment is about type validation, however (i.e., make sure the values submitted are the correct types based on the pipeline signature), not about the type annotation. Can you re-add the comment?

droctothorpe replied:
Added the comments back! Thanks, @connor-mccarthy!

connor-mccarthy replied:
Can you also update the title since the comments aren't removed? Also, since the TODOs are unrelated, this is a fairly small PR. Are there other needed `client.py` type annotation updates that we can bundle in to this PR?

droctothorpe replied:
Good call. Happy to tackle that.

droctothorpe replied:
Updated about 20 type hints and modified the commit and PR title. Happy to address any other feedback.

connor-mccarthy reviewed (commented):
One question for you

connor-mccarthy commented on sdk/python/kfp/client/client.py:
```diff
@@ -303,7 +303,7 @@ def _load_config(
         return config
 
-    def _is_inverse_proxy_host(self, host):
+    def _is_inverse_proxy_host(self, host: str):
```
nit: return annotation

Can you scan for this throughout?

droctothorpe replied:
Done! 👍

connor-mccarthy commented on sdk/python/kfp/client/client.py:
```diff
@@ -694,9 +694,9 @@ def _override_caching_options(self, pipeline_obj: dict,
-        sort_by: str = '',
+        page_token: Optional[str] = '',
+        page_size: Optional[int] = 10,
+        sort_by: Optional[str] = '',
```
`typing.Optional[<type>]` == `Union[<type>, None]` [[ref](https://docs.python.org/3/library/typing.html#typing.Optional)], which is not the same as a parameter having a default value.

Since the default is not `None` and (as far as I know) we don't permit passing `None`, the original annotation is correct.

droctothorpe replied:
TIL. Thanks! I wonder if there is a way to specify optional-ness but exclude None, because it is optional from a UX standpoint, but maybe that doesn't matter?

connor-mccarthy reviewed (changes requested):
Can you also run `mypy` over this file? It's not critical to resolve every `mypy` error message,  but a before/after `mypy` output comparison is a good way to assess how much progress is made and if there is other low-hanging fruit to further reconcile inconsistencies.

droctothorpe commented:
> > Can you also run `mypy` over this file? It's not critical to resolve every `mypy` error message, but a before/after `mypy` output comparison is a good way to assess how much progress is made and if there is other low-hanging fruit to further reconcile inconsistencies.
> 
> Copy that. Thanks, Connor. Plan to tackle on Monday.

Scratch that, heh. I don't know how to stop working. Addressed feedback. A few minor questions that I'll ask alongside the relevant code.

droctothorpe commented on sdk/python/kfp/client/client.py:
```diff
@@ -1057,7 +1057,10 @@ def create_run_from_pipeline_func(
         """
         #TODO: Check arguments against the pipeline function
-        pipeline_name = pipeline_func.name
+        if isinstance(pipeline_func, base_component.BaseComponent):
+            pipeline_name = pipeline_func.name
+        else:
+            pipeline_name = pipeline_func.__name__
```
This one is a bit odd. The [type hint](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/client/client.py#L1026) specifies a union:
> pipeline_func: Union[Callable[..., Any], base_component.BaseComponent],

`BaseComponent` has a `name` attribute. A generic callable doesn't. Which is why mypy complained about this. Not sure this is the optimal solution and totally open to alternatives.

connor-mccarthy replied:
In this case, I believe the annotation should be `base_component.BaseComponent` only.

The `Callable` annotation is a relic from v1, where pipelines were pure functions. In v2, pipelines created with the `@dsl.pipeline` decorator are instances of `GraphComponent`, which is a subclass of `BaseComponent`.

connor-mccarthy replied:
In this case, I believe the annotation should be `base_component.BaseComponent` only.

The `Callable` annotation is a relic from v1, where pipelines were pure functions. In v2, pipelines created with the `@dsl.pipeline` decorator are instances of `GraphComponent`, which is a subclass of `BaseComponent`.

droctothorpe commented on sdk/python/kfp/client/client.py:
```diff
@@ -303,9 +303,9 @@ def _load_config(
+    def _is_inverse_proxy_host(self, host: str) -> bool:
         if host:
-            return re.match(r'\S+.googleusercontent.com/{0,1}$', host)
+            return bool(re.match(r'\S+.googleusercontent.com/{0,1}$', host))
```
`re.match` does not return a bool, but it does return a truthy object if a match is found, and `None` (falsey) if a match is not found. Let me know if you want to be more explicit about this, e.g.:
```
if re.match(r'\S+.googleusercontent.com/{0,1}$', host) is None:
    return False
```

connor-mccarthy replied:
Perhaps it makes sense to expand with an explicit `None` check, since it does not require future readers to know about the truthiness/falsiness properties of `re.Match` objects.

droctothorpe commented on sdk/python/kfp/client/client.py:
```diff
@@ -303,14 +303,10 @@ def _load_config(
+    def _is_inverse_proxy_host(self, host: str) -> bool:
         if host:
-            return re.match(r'\S+.googleusercontent.com/{0,1}$', host)
-        if re.match(r'\w+', host):
-            warnings.warn(
-                f'The received host is {host}, please include the full endpoint'
-                ' address (with ".(pipelines/notebooks).googleusercontent.com")'
-            )
```
AFACT, this code is never reached. The only time that we reach line 309 is when `host` is `""` or `None`. Since both are falsey, this condition never evaluates as `True`. As a result, the subsequent warning on lines 310-313 is never executed.

We may want to handle `host` being `""` or `None` with a simple `else` aligned to the `if host` conditional, but the warning should probably be reworded in that case.

connor-mccarthy replied:
Casting `re.Match` or `None` to a boolean and `-> bool` SGTM.

droctothorpe commented on sdk/python/kfp/client/client.py:
```diff
@@ -303,14 +303,10 @@ def _load_config(
-                f'The received host is {host}, please include the full endpoint'
-                ' address (with ".(pipelines/notebooks).googleusercontent.com")'
-            )
+            if re.match(r'\S+.googleusercontent.com/{0,1}$', host) is not None:
```
Used the negative to consolidate the `return False` statements into one, i.e. we return `False` if the regex doesn't match _or_ if the host is falsey.

droctothorpe commented on sdk/python/kfp/client/client.py:
```diff
@@ -1362,9 +1358,9 @@ def _get_workflow_json(self, run_id: str) -> dict:
-        pipeline_package_path: str = None,
-        pipeline_name: str = None,
-        description: str = None,
+        pipeline_package_path: str,
```
```
❯ python -u "/Users/inn487/code/kubeflow/pipelines/test_args.py"
/Users/inn487/code/kubeflow/pipelines/sdk/python/kfp/client/client.py:141: FutureWarning: This client only works with Kubeflow Pipeline v2.0.0-alpha.0 and later versions.
  warnings.warn(
Traceback (most recent call last):
  File "/Users/inn487/code/kubeflow/pipelines/test_args.py", line 53, in <module>
    print(client.upload_pipeline())
  File "/Users/inn487/code/kubeflow/pipelines/sdk/python/kfp/client/client.py", line 1378, in upload_pipeline
    response = self._upload_api.upload_pipeline(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/kfp-upstream/lib/python3.8/site-packages/kfp_server_api/api/pipeline_upload_service_api.py", line 69, in upload_pipeline
    return self.upload_pipeline_with_http_info(uploadfile, **kwargs)  # noqa: E501
  File "/opt/homebrew/Caskroom/miniconda/base/envs/kfp-upstream/lib/python3.8/site-packages/kfp_server_api/api/pipeline_upload_service_api.py", line 132, in upload_pipeline_with_http_info
    raise ApiValueError("Missing the required parameter `uploadfile` when calling `upload_pipeline`")  # noqa: E501
kfp_server_api.exceptions.ApiValueError: Missing the required parameter `uploadfile` when calling `upload_pipeline`
```

connor-mccarthy reviewed (approved):
/lgtm
/approve

Thanks for this, @droctothorpe!

connor-mccarthy commented:
/ok-to-test

connor-mccarthy commented:
Manually merging, as `kubeflow-pipelines-sdk-execution-tests` failures are unrelated to these changes.
